### ISSUE TITLE
[bug] CMakeToolchain changes the value of string variables unexpectedly/incorrectly

### ISSUE THREAD

**Issue by phillip-keldenich** [conan-io/conan#16315]
### Describe the bug

Linux (Ubuntu 22.04.4 LTS)
python 3.8.12 (Anaconda)
conan 2.3.1

I have the following code as part of a conan package that downloads and builds C++ bindings for a popular commercial MIP solver (Gurobi).

```
from conan.tools.cmake import CMakeToolchain, CMake, cmake_layout

[...]

def generate(self):
        dl_result = self._download_and_extract()
        c_lib_location = dl_result["lib"]
        c_imp_location = dl_result["imp"]
        target_dir = dl_result["dir"]
        tc = CMakeToolchain(self)
        td = target_dir.replace("\\", "/")
        tc.variables["GUROBI_EXTRACT_TARGET_DIR"] = td
        tc.variables["GUROBI_C_SHARED_LIB_LOCATION"] = (
            td + "/" + c_lib_location.replace("\\", "/")
        )
        if c_imp_location:
            tc.variables["GUROBI_C_SHARED_IMP_LOCATION"] = (
                td + "/" + c_imp_location.replace("\\", "/")
            )
        if self.settings.get_safe("os") == "Windows":
            tc.variables["GUROBI_FPIC"] = "Off"
        else:
            tc.variables["GUROBI_FPIC"] = (
                "On" if self.options.get_safe("fPIC") else "Off"
            )
        tc.variables["GUROBI_VERSION"] = self.version
        tc.generate()
```

In the generate method, the appropriate pre-built binaries for the target arch, build type, etc. are downloaded automatically and extracted; this returns a set of paths to various components of the MIP solver (which depend on the platform).
These paths are passed to the CMakeLists.txt I created for building the C++ bindings of the MIP solver (for which source code is included in the downloaded/extracted bundle) using the `CMakeToolchain` `variables` map.

The package builds fine on all systems with conventional setups; however, on a certain machine, the build fails.
The paths all appear to be correct on the python side (I added print(...) statements and manually inspected the values, they point to the right location and the files all exist).
I believe that I have traced it down to some processing of the string variables that happens when conan generates the `conan_toolchain.cmake` file.

For example, if I print the value `td` given to the variable `GUROBI_EXTRACT_TARGET_DIR`, on the machine in question I get

`/misc/ibr/home/my_username/.conan2/p/b/gurob4ddde3d219d31/b/build/Release/generators`, 

which is the expected path (the extraction process pulls a few selected files and subfolders from the downloaded bundle and places them in this folder). On the other hand, the generated conan_toolchain.cmake file reflects this variable setting with the line

```
set(GUROBI_EXTRACT_TARGET_DIR "/misc${CMAKE_CURRENT_LIST_DIR}/../../../build/Release/generators" CACHE STRING "Variable GUROBI_EXTRACT_TARGET_DIR conan-toolchain defined")
```

This is not correct; it expands to 

`/misc/misc/ibr/home/my_username/.conan2/p/b/gurob4ddde3d219d31/b/build/Release/generators`

(note the double `misc`), which does not point to an existing file.

Note: it is not impossible that this is due to the fact that my home directory (which I build in) is reachable by two different paths (`/misc/ibr/home/my_username` and `/ibr/home/my_username`) via a symlink I have no control over.

Overall, I find this type of postprocessing of string values quite dangerous and unexpected and I would have expected the values to be passed to CMake without change or interpretation (after all, they might not even be paths); however, I could see that there might be some cases that require it.
For now, I can probably work around the issue by doing some crude replacement of a part of the path passed to CMake with another variable (i.e., sneak the path past conans processing, e.g., by replacing all slashes with a `${VAR_THAT_CONTAINS_A_SLASH}` or similar).

### How to reproduce it

I doubt you can reproduce the erroneous behavior easily; it seems to work on most platforms and I really believe it has to do with the slightly non-standard home directory which contains symlinks in the path to it.
To make a minimum non-working example, one could probably eliminate the downloading/extraction part and anything related to the actual MIP solver and just attempt to pass a path containing the value of `${CMAKE_CURRENT_LIST_DIR}/some/dir` in `conan_toolchain.cmake` to a `CMakeLists.txt` via `tc.variables["TEST_VAR"] = some_path`.

**Comment by phillip-keldenich** [comment]
I checked the workaround:

```
        tc = CMakeToolchain(self)
        td = target_dir.replace("\\", "/")
        td = td.replace("/", "${VAR_THAT_CONTAINS_A_SLASH}")
        tc.variables["VAR_THAT_CONTAINS_A_SLASH"] = "/"
        tc.variables["GUROBI_EXTRACT_TARGET_DIR"] = td
```

and it results in a working build; of course, this is not nice and raises the question whether the order of the 'set(...)' commands in `conan_toolchain.cmake` always reflects the order of setting the values on the python side (because the workaround would break if `VAR_THAT_CONTAINS_A_SLASH` was set after `GUROBI_EXTRACT_TARGET_DIR`.

**Comment by memsharded** [comment]
Hi @phillip-keldenich 

Thanks for your report.

It seems the issue is that the relativization of paths if finding ``/home/my_username/.conan2/p/b/gurob4ddde3d219d31`` as a valid "full path", but then it happens that it is located as subfolder in a very non conventional file system that starts with ``/misc/ibr`` and that was not expected. I'll have a look to see if this can be fixed.

**Comment by memsharded** [comment]
> Note: it is not impossible that this is due to the fact that my home directory (which I build in) is reachable by two different paths (/misc/ibr/home/my_username and /ibr/home/my_username) via a symlink I have no control over.

This is indeed what is confusing the system, and the reason it doesn't happen in the vast majority of other cases

> Overall, I find this type of postprocessing of string values quite dangerous and unexpected and I would have expected the values to be passed to CMake without change or interpretation (after all, they might not even be paths); however, I could see that there might be some cases that require it.

We have been improving this and adding explicit relativization of things when they are known to be paths. The original reason was the demand to make generated files relocatable (see https://blog.conan.io/2023/05/23/Conan-agnostic-deploy-dependencies.html), but it is true that the current replacement strategy is not great, I will try to do another iteration over this.

**Comment by phillip-keldenich** [comment]
Thanks for the quick response/reaction! I had something like this in mind when I said I could see reasons one would like to do a replacement like this. For me, this issue now has little time pressure/sensitivity since I have a workaround (albeit one that might be a bit brittle).

**Comment by memsharded** [comment]
I am proposing https://github.com/conan-io/conan/pull/16316 for next Conan 2.4 release, in case you want to give it a try and run it from sources from my branch with your code.

> Thanks for the quick response/reaction! I had something like this in mind when I said I could see reasons one would like to do a replacement like this. For me, this issue now has little time pressure/sensitivity since I have a workaround (albeit one that might be a bit brittle).

Thanks very much for your understanding and consideration, it is really appreciated. Very often maintainers of very active open source projects are subject to lots of pressure from users to fix things and are overwhelmed with demands, so this understanding is great.

But I agree that we should be pursuing correctness and it is not the first time someone is bitten by this raw replacement, so I am pushing the PR to get rid of it, and do a better more explicit replacements of things known to be paths. It is possible that this might need some further iteration, because some missing relativizations, but lets wait for users feedback about it.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/16316 was merged for next 2.4
